cpp/src/buffer/buffer.cpp

```diff
@@ -99,4 +103,120 @@ std::unique_ptr<Buffer> Buffer::copy(MemoryType target, rmm::cuda_stream_view st
     );
 }
 
+std::unique_ptr<Buffer> Buffer::copy_slice(
+    std::ptrdiff_t offset, std::ptrdiff_t length, rmm::cuda_stream_view stream
+) const {
+    RAPIDSMPF_EXPECTS(
+        offset <= std::ptrdiff_t(size), "offset can't be greater than size"
+    );
+    RAPIDSMPF_EXPECTS(
+        offset + length <= std::ptrdiff_t(size),
+        "offset + length can't be greater than size"
+    );
+    return std::visit(
+        overloaded{
+            [&](const HostStorageT& storage) {
+                return std::unique_ptr<Buffer>(new Buffer{
+                    std::make_unique<std::vector<uint8_t>>(
+                        storage->begin() + offset, storage->begin() + offset + length
+                    ),
+                    br
+                });
+            },
+            [&](DeviceStorageT const& storage) {
+                return std::unique_ptr<Buffer>(new Buffer{
+                    std::make_unique<rmm::device_buffer>(
+                        static_cast<cuda::std::byte*>(storage->data()) + offset,
+                        length,
+                        stream,
+                        br->device_mr()
+                    ),
+                    br
+                });
+            }
+        },
+        storage_
+    );
+}
+
+std::unique_ptr<Buffer> Buffer::copy_slice(
+    MemoryType target,
+    std::ptrdiff_t offset,
+    std::ptrdiff_t length,
+    rmm::cuda_stream_view stream
+) const {
+    RAPIDSMPF_EXPECTS(
+        offset <= std::ptrdiff_t(size), "offset can't be greater than size"
+    );
+    RAPIDSMPF_EXPECTS(
+        offset + length <= std::ptrdiff_t(size),
+        "offset + length can't be greater than size"
+    );
+
+    if (mem_type() == target) {
+        return copy_slice(offset, length, stream);
```

Comment:
*nit*: This is why I don't really like these "C-like" APIs. It's very easy to switch up `offset` and `length` since they are the same type.